File: backend/app/routers/vehicles.py
import uuid
from pathlib import Path
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import Vehicle
from ..schemas import VehicleCreate, VehicleUpdate, VehicleOut
from ..deps import get_current_user
from ..config import settings
from ..models import User

router = APIRouter(prefix="/vehicles", tags=["vehicles"])

ALLOWED_IMAGE = {"image/jpeg", "image/png", "image/gif", "image/webp"}


@router.get("", response_model=list[VehicleOut])
def list_vehicles(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return db.query(Vehicle).order_by(Vehicle.year.desc(), Vehicle.make).all()


@router.post("", response_model=VehicleOut, status_code=status.HTTP_201_CREATED)
def create_vehicle(
    data: VehicleCreate,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    v = Vehicle(**data.model_dump())
    db.add(v)
    db.commit()
    db.refresh(v)
    return v


@router.get("/{vehicle_id}", response_model=VehicleOut)
def get_vehicle(
    vehicle_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    v = db.query(Vehicle).filter(Vehicle.id == vehicle_id).first()
    if not v:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    return v


@router.patch("/{vehicle_id}", response_model=VehicleOut)
def update_vehicle(
    vehicle_id: int,
    data: VehicleUpdate,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    v = db.query(Vehicle).filter(Vehicle.id == vehicle_id).first()
    if not v:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    for k, val in data.model_dump(exclude_unset=True).items():
        setattr(v, k, val)
    db.commit()
    db.refresh(v)
    return v


@router.delete("/{vehicle_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_vehicle(
    vehicle_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    v = db.query(Vehicle).filter(Vehicle.id == vehicle_id).first()
    if not v:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    db.delete(v)
    db.commit()
    return None


@router.post("/{vehicle_id}/photo", response_model=VehicleOut)
async def upload_vehicle_photo(
    vehicle_id: int,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    if file.content_type not in ALLOWED_IMAGE:
        raise HTTPException(status_code=400, detail="File must be JPEG, PNG, GIF, or WebP")
    v = db.query(Vehicle).filter(Vehicle.id == vehicle_id).first()
    if not v:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    ext = Path(file.filename or "img").suffix or ".jpg"
    name = f"{vehicle_id}_{uuid.uuid4().hex[:8]}{ext}"
    out_dir = settings.upload_dir / "vehicles"
    out_dir.mkdir(parents=True, exist_ok=True)
    path = out_dir / name
    content = await file.read()
    path.write_bytes(content)
    v.photo_path = f"vehicles/{name}"
    db.commit()
    db.refresh(v)
    return v

File: frontend/src/api/vehicles.ts
import { api, apiFormData, uploadsUrl } from "./client";

export interface Vehicle {
  id: number;
  nickname: string | null;
  make: string;
  model: string;
  trim: string | null;
  year: number;
  vin: string | null;
  license_plate: string | null;
  current_mileage: number | null;
  photo_path: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface VehicleCreate {
  nickname?: string | null;
  make: string;
  model: string;
  trim?: string | null;
  year: number;
  vin?: string | null;
  license_plate?: string | null;
  current_mileage?: number | null;
  photo_path?: string | null;
}

export interface VehicleUpdate {
  nickname?: string | null;
  make?: string;
  model?: string;
  trim?: string | null;
  year?: number;
  vin?: string | null;
  license_plate?: string | null;
  current_mileage?: number | null;
  photo_path?: string | null;
}

export async function listVehicles(): Promise<Vehicle[]> {
  return api<Vehicle[]>("/vehicles");
}

export async function getVehicle(id: number): Promise<Vehicle> {
  return api<Vehicle>(`/vehicles/${id}`);
}

export async function createVehicle(data: VehicleCreate): Promise<Vehicle> {
  return api<Vehicle>("/vehicles", { method: "POST", body: JSON.stringify(data) });
}

export async function updateVehicle(id: number, data: VehicleUpdate): Promise<Vehicle> {
  return api<Vehicle>(`/vehicles/${id}`, { method: "PATCH", body: JSON.stringify(data) });
}

export async function deleteVehicle(id: number): Promise<void> {
  return api<void>(`/vehicles/${id}`, { method: "DELETE" });
}

export async function uploadVehiclePhoto(vehicleId: number, file: File): Promise<Vehicle> {
  const form = new FormData();
  form.append("file", file);
  return apiFormData<Vehicle>(`/vehicles/${vehicleId}/photo`, form);
}

export { uploadsUrl };
